File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/screenshots/TestMain.java
package org.totschnig.myexpenses.test.screenshots;

import android.Manifest;
import android.content.Intent;
import android.content.SharedPreferences;

import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;
import org.totschnig.myexpenses.BuildConfig;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.MyExpenses;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.testutils.BaseUiTest;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.util.distrib.DistributionHelper;

import java.util.Locale;
import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.test.core.app.ActivityScenario;
import androidx.test.espresso.Espresso;
import androidx.test.espresso.NoMatchingViewException;
import androidx.test.espresso.ViewAction;
import androidx.test.espresso.contrib.DrawerActions;
import androidx.test.espresso.contrib.RecyclerViewActions;
import androidx.test.rule.GrantPermissionRule;
import tools.fastlane.screengrab.Screengrab;
import tools.fastlane.screengrab.locale.LocaleTestRule;
import tools.fastlane.screengrab.locale.LocaleUtil;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.matcher.ViewMatchers.hasDescendant;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withSubstring;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.containsString;
import static org.hamcrest.Matchers.instanceOf;
import static org.totschnig.myexpenses.testutils.Matchers.first;

/**
 * This test is meant to be run with FastLane Screengrab, but also works on its own.
 */
public class TestMain extends BaseUiTest {
  @ClassRule
  public static final LocaleTestRule localeTestRule = new LocaleTestRule();
  private ActivityScenario<MyExpenses> activityScenario = null;
  @Rule
  public final GrantPermissionRule grantPermissionRule = GrantPermissionRule.grant(
      Manifest.permission.WRITE_CALENDAR, Manifest.permission.READ_CALENDAR);

  @Test
  public void mkScreenShots() {
    loadFixture(BuildConfig.TEST_SCENARIO == 2);
    scenario();
  }

  private void drawerAction(ViewAction action) {
    //no drawer on w700dp
    try {
      onView(withId(R.id.drawer)).perform(action);
    } catch (NoMatchingViewException ignored) { }
  }

  private void scenario() {
    sleep();
    switch (BuildConfig.TEST_SCENARIO) {
      case 1: {
        drawerAction(DrawerActions.open());
        takeScreenshot("summarize");
        drawerAction(DrawerActions.close());
        takeScreenshot("group");
        clickMenuItem(R.id.RESET_COMMAND);
        Espresso.closeSoftKeyboard();
        takeScreenshot("export");
        Espresso.pressBack();
        onView(withSubstring(getString(R.string.split_transaction))).perform(click());
        onView(withId(android.R.id.button1)).perform(click());
        Espresso.closeSoftKeyboard();
        takeScreenshot("split");
        Espresso.pressBack();
        clickMenuItem(R.id.DISTRIBUTION_COMMAND);
        takeScreenshot("distribution");
        Espresso.pressBack();
        clickMenuItem(R.id.HISTORY_COMMAND);
        clickMenuItem(R.id.GROUPING_COMMAND);
        onView(withText(R.string.grouping_month)).perform(click());
        clickMenuItem(R.id.TOGGLE_INCLUDE_TRANSFERS_COMMAND);
        takeScreenshot("history");
        Espresso.pressBack();
        clickMenuItem(R.id.BUDGET_COMMAND);
        onView(withId(R.id.recycler_view))
            .perform(RecyclerViewActions.actionOnItemAtPosition(0, click()));
        takeScreenshot("budget");
        Espresso.pressBack();
        Espresso.pressBack();
        clickMenuItem(R.id.SETTINGS_COMMAND);
        onView(instanceOf(RecyclerView.class))
            .perform(RecyclerViewActions.actionOnItem(hasDescendant(withText(R.string.synchronization)),
                click()));
        onView(instanceOf(RecyclerView.class))
            .perform(RecyclerViewActions.actionOnItem(hasDescendant(withText(R.string.pref_manage_sync_backends_title)),
                click()));
        onView(withText(containsString("Drive"))).perform(click());
        onView(withText(containsString("Dropbox"))).perform(click());
        onView(withText(containsString("WebDAV"))).perform(click());
        takeScreenshot("sync");
        break;
      }
      case 2: {//tablet screenshots
        takeScreenshot("main");
        clickMenuItem(R.id.DISTRIBUTION_COMMAND);
        takeScreenshot("distribution");
        Espresso.pressBack();

        onView(first(withText(containsString(getTestContext().getString(org.totschnig.myexpenses.debug.test.R.string.testData_transaction1SubCat))))).perform(click());
        onView(withId(android.R.id.button1)).perform(click());
        Espresso.pressBack();//close keyboard
        onView(withId(R.id.PictureContainer)).perform(click());
        takeScreenshot("edit");
        break;
      }
      default: {
        throw new IllegalArgumentException("Unknown scenario" + BuildConfig.TEST_SCENARIO);
      }
    }

  }

  private void loadFixture(@SuppressWarnings("SameParameterValue") boolean withPicture) {
    //LocaleTestRule only configure for app context, fixture loads resources from instrumentation context
    final Locale testLocale = LocaleUtil.getTestLocale();
    if (testLocale != null) {//if run from Android Studio and not via Screengrab
      configureLocale(testLocale);
    }
    SharedPreferences pref = getApp().getSettings();
    if (pref == null)
      Assert.fail("Could not find prefs");
    pref.edit().putString(PrefKey.HOME_CURRENCY.getKey(), Utils.getSaveDefault().getCurrencyCode()).apply();
    getApp().getLicenceHandler().setLockState(false);

    getApp().fixture.setup(withPicture);
    int current_version = DistributionHelper.getVersionNumber();
    pref.edit()
        .putLong(PrefKey.CURRENT_ACCOUNT.getKey(), getApp().fixture.getAccount1().getId())
        .putInt(PrefKey.CURRENT_VERSION.getKey(), current_version)
        .putInt(PrefKey.FIRST_INSTALL_VERSION.getKey(), current_version)
        .apply();
    final Intent startIntent = new Intent(getApp(), MyExpenses.class);
    activityScenario = ActivityScenario.launch(startIntent);
  }

  private void sleep() {
    try {
      Thread.sleep(500);
    } catch (InterruptedException e) {
      e.printStackTrace();
    }
  }

  private void takeScreenshot(String fileName) {
    Espresso.onIdle();
    Screengrab.screenshot(fileName);
  }

  @NonNull
  @Override
  protected ActivityScenario<? extends ProtectedFragmentActivity> getTestScenario() {
    return Objects.requireNonNull(activityScenario);
  }
}

File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/screenshots/TestMain.kt
package org.totschnig.myexpenses.test.screenshots

import android.Manifest
import android.content.res.Configuration
import android.content.res.Resources
import androidx.appcompat.app.AppCompatDelegate
import androidx.compose.ui.test.assertTextContains
import androidx.compose.ui.test.filter
import androidx.compose.ui.test.hasText
import androidx.compose.ui.test.onChildren
import androidx.compose.ui.test.onFirst
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.core.os.LocaleListCompat
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.closeSoftKeyboard
import androidx.test.espresso.Espresso.onIdle
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.Espresso.pressBack
import androidx.test.espresso.NoMatchingViewException
import androidx.test.espresso.ViewAction
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.scrollTo
import androidx.test.espresso.contrib.DrawerActions
import androidx.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import androidx.test.rule.GrantPermissionRule
import androidx.test.uiautomator.UiDevice
import org.hamcrest.Matchers.containsString
import org.junit.After
import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.Rule
import org.totschnig.myexpenses.BuildConfig
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.test.espresso.SettingsTest
import org.totschnig.myexpenses.testutils.BaseMyExpensesTest
import org.totschnig.myexpenses.testutils.withPositionInParent
import org.totschnig.myexpenses.util.distrib.DistributionHelper.versionNumber
import tools.fastlane.screengrab.Screengrab
import tools.fastlane.screengrab.cleanstatusbar.CleanStatusBar
import tools.fastlane.screengrab.locale.LocaleUtil

abstract class TestMain : BaseMyExpensesTest() {

    open val shouldTakeScreenShot = false

    @get:Rule
    val grantPermissionRule: GrantPermissionRule = GrantPermissionRule.grant(
        Manifest.permission.WRITE_CALENDAR, Manifest.permission.READ_CALENDAR
    )

    @After
    fun cleanUp() {
        app.fixture.cleanup(contentResolver)
    }

    fun runScenario(scenario: String, locale: String?) {
        loadFixture(scenario == "2", locale)
        scenario(scenario)
        testScenario.onActivity {
            it.runOnUiThread {
                AppCompatDelegate.setApplicationLocales(LocaleListCompat.getEmptyLocaleList())
            }
        }
    }

    private fun drawerAction(action: ViewAction) {
        //no drawer on w700dp
        try {
            onView(withId(R.id.drawer)).perform(action)
        } catch (ignored: NoMatchingViewException) {
        }
    }

    private fun scenario(scenario: String) {
        when (scenario) {
            "1" -> {
                drawerAction(DrawerActions.open())
                takeScreenshot("summarize")
                drawerAction(DrawerActions.close())
                takeScreenshot("group")
                clickMenuItem(R.id.RESET_COMMAND)
                closeSoftKeyboard()
                takeScreenshot("export")
                pressBack()
                listNode.onChildren().onFirst()
                    .assertTextContains(getString(R.string.split_transaction), substring = true)
                clickContextItem(R.string.details)
                onView(withId(android.R.id.button1)).perform(click())
                closeSoftKeyboard()
                takeScreenshot("split")
                pressBack()
                clickMenuItem(R.id.DISTRIBUTION_COMMAND)
                takeScreenshot("distribution")
                pressBack()
                clickMenuItem(R.id.HISTORY_COMMAND)
                clickMenuItem(R.id.GROUPING_COMMAND)
                onView(withText(R.string.grouping_month)).perform(click())
                clickMenuItem(R.id.TOGGLE_INCLUDE_TRANSFERS_COMMAND)
                takeScreenshot("history")
                pressBack()
                clickMenuItem(R.id.BUDGET_COMMAND)
                onView(withId(R.id.recycler_view))
                    .perform(actionOnItemAtPosition<RecyclerView.ViewHolder>(0, click()))
                val device = UiDevice.getInstance(getInstrumentation())
                device.setOrientationRight()
                onIdle()
                //wait for sum to load IdlingResource is too cumbersome to set up, since
                //onActivity does not get us hold on BudgetActivity
                Thread.sleep(500)
                takeScreenshot("budget")
                device.setOrientationNatural()
                onIdle()
                Thread.sleep(500)
                pressBack()
                pressBack()
                clickMenuItem(R.id.SETTINGS_COMMAND)
                SettingsTest.navigateTo(
                    R.string.synchronization,
                    R.string.pref_manage_sync_backends_title
                )
                onView(withText(containsString("Drive"))).perform(click())
                onView(withText(containsString("Dropbox"))).perform(click())
                onView(withText(containsString("WebDAV"))).perform(scrollTo(), click())
                Thread.sleep(5000)
                takeScreenshot("sync")
            }

            "2" -> {
                //tablet screenshots
                takeScreenshot("main")
                clickMenuItem(R.id.DISTRIBUTION_COMMAND)
                takeScreenshot("distribution")
                pressBack()
                listNode.onChildren()
                    .filter(
                        hasText(
                            testContext.getString(org.totschnig.myexpenses.test.R.string.testData_transaction1SubCat),
                            substring = true
                        )
                    )
                    .onFirst()
                    .performClick()
                composeTestRule.onNodeWithText(getString(R.string.menu_edit)).performClick()
                pressBack() //close keyboard
                onView(withPositionInParent(R.id.AttachmentGroup, 0))
                    .perform(click())
                takeScreenshot("edit")
            }

            else -> {
                throw IllegalArgumentException("Unknown scenario" + BuildConfig.TEST_SCENARIO)
            }
        }
    }

    private fun loadFixture(withPicture: Boolean, locale: String?) {
        //LocaleTestRule only configure for app context, fixture loads resources from instrumentation context
        LocaleUtil.localeFromString(locale)?.let {
            AppCompatDelegate.setApplicationLocales(LocaleListCompat.create(it))
            val config = Configuration()
            config.locale = it
            testContext.resources.update(config)
            targetContext.resources.update(config)
        }
        unlock()
        app.fixture.setup(withPicture, repository, app.appComponent.plannerUtils(), homeCurrency)
        prefHandler.putLong(PrefKey.CURRENT_ACCOUNT, app.fixture.account1.id)
        prefHandler.putInt(PrefKey.CURRENT_VERSION, versionNumber)
        prefHandler.putInt(PrefKey.FIRST_INSTALL_VERSION, versionNumber)
        launch()
    }

    private fun Resources.update(configuration: Configuration) {
        updateConfiguration(configuration, displayMetrics)
    }


    private fun takeScreenshot(fileName: String) {
        if (shouldTakeScreenShot) {
            onIdle()
            Screengrab.screenshot(fileName)
        }
    }

    companion object {

        @JvmStatic
        @BeforeClass
        fun beforeAll() {
            CleanStatusBar.enableWithDefaults()
        }

        @JvmStatic
        @AfterClass
        fun afterAll() {
            CleanStatusBar.disable()
        }
    }
}